File: java/src/main/java/io/aurigraph/dlt/sdk/namespace/NodesApi.java
package io.aurigraph.dlt.sdk.namespace;

import io.aurigraph.dlt.sdk.AurigraphClient;
import io.aurigraph.dlt.sdk.dto.NodeInfo;
import io.aurigraph.dlt.sdk.dto.NodeList;
import io.aurigraph.dlt.sdk.dto.NodeMetrics;
import io.aurigraph.dlt.sdk.dto.NodeRegisterRequest;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

/**
 * Namespace for /api/v11/nodes endpoints.
 */
public class NodesApi {

    private final AurigraphClient client;

    public NodesApi(AurigraphClient client) {
        this.client = client;
    }

    public NodeList list() {
        return list(0, 50);
    }

    public NodeList list(int page, int pageSize) {
        return client.get("/nodes?page=" + page + "&pageSize=" + pageSize, NodeList.class);
    }

    public NodeInfo get(String nodeId) {
        String enc = URLEncoder.encode(nodeId, StandardCharsets.UTF_8);
        return client.get("/nodes/" + enc, NodeInfo.class);
    }

    public NodeInfo register(NodeRegisterRequest request) {
        return client.post("/nodes", request, NodeInfo.class);
    }

    public NodeMetrics getMetrics() {
        return client.get("/nodes/metrics", NodeMetrics.class);
    }
}

File: android/aurigraph-sdk/src/main/kotlin/io/aurigraph/sdk/apis/NodesApi.kt
package io.aurigraph.sdk.apis

import io.aurigraph.sdk.AurigraphClient
import io.aurigraph.sdk.models.NodeInfo
import io.aurigraph.sdk.models.NodeMetrics
import io.aurigraph.sdk.models.NodeRegisterRequest
import kotlinx.serialization.json.JsonObject

/**
 * Network node management -- listing, registration, and metrics.
 */
class NodesApi(private val client: AurigraphClient) {

    /** List all nodes in the network. */
    suspend fun list(page: Int = 0, pageSize: Int = 50): JsonObject =
        client.get("/nodes?page=$page&pageSize=$pageSize")

    /** Get a specific node by ID. */
    suspend fun get(nodeId: String): NodeInfo = client.get("/nodes/$nodeId")

    /** Get aggregate node metrics (total, active, validators, network status). */
    suspend fun metrics(): NodeMetrics = client.get("/nodes/metrics")

    /** Get node storage stats. */
    suspend fun stats(): JsonObject = client.get("/nodes/stats")

    /** Register a new node in the network. */
    suspend fun register(request: NodeRegisterRequest): NodeInfo =
        client.post("/nodes", request)
}
